crates/biome_js_analyze/src/services/semantic_class.rs

```diff
@@ -1,25 +1,149 @@
 use biome_js_syntax::{
-    AnyJsClassMember, AnyJsExpression, JsArrayAssignmentPattern, JsArrowFunctionExpression,
-    JsAssignmentExpression, JsClassMemberList, JsConstructorClassMember, JsFunctionBody,
-    JsLanguage, JsObjectAssignmentPattern, JsObjectBindingPattern, JsPostUpdateExpression,
-    JsPreUpdateExpression, JsPropertyClassMember, JsStaticMemberAssignment,
-    JsStaticMemberExpression, JsSyntaxKind, JsSyntaxNode, JsVariableDeclarator, TextRange,
-    TsPropertyParameter,
+    AnyJsClassMember, AnyJsExpression, AnyJsRoot, JsArrayAssignmentPattern,
+    JsArrowFunctionExpression, JsAssignmentExpression, JsClassDeclaration, JsClassMemberList,
+    JsConstructorClassMember, JsFunctionBody, JsLanguage, JsObjectAssignmentPattern,
+    JsObjectBindingPattern, JsPostUpdateExpression, JsPreUpdateExpression, JsPropertyClassMember,
+    JsStaticMemberAssignment, JsStaticMemberExpression, JsSyntaxKind, JsSyntaxNode,
+    JsVariableDeclarator, TextRange, TsPropertyParameter,
 };
 
-use biome_analyze::QueryMatch;
+use biome_analyze::{
+    AddVisitor, FromServices, Phase, Phases, QueryKey, QueryMatch, Queryable, RuleKey,
+    RuleMetadata, ServiceBag, ServicesDiagnostic, Visitor, VisitorContext, VisitorFinishContext,
+};
 use biome_rowan::{
     AstNode, AstNodeList, AstSeparatedList, SyntaxNode, Text, WalkEvent, declare_node_union,
 };
 use std::collections::HashSet;
 
+#[derive(Clone)]
+pub struct SemanticClassServices {
+    class_member_references: ClassMemberReferences,
+}
+
+impl SemanticClassServices {
+    pub fn class_member_references(&self) -> &ClassMemberReferences {
+        &self.class_member_references
+    }
+}
```

Comment:
Since we're moving everything inside a service, you need to think about how rules are going to query it. The method `class_member_references` is fine, but as I highlighted before, **the service is a singleton for the entire CST root**, and if we have multiple classes, it needs to query the references of the correct `class`. Example of how the rule would do it:

```rs
let node = ctx.query();
let class_references = ctx.class_member_references(node.syntax().trimmed_range());
```

Do you think it makes sense?